skyportal/models.py

```diff
@@ -183,19 +183,66 @@ def desi_dr8_url(self):
                 f"&dec={self.dec}&size=200&layer=dr8&pixscale=0.262&bands=grz")
 
 
-GroupSource = join_model('group_sources', Group, Source)
+class Filter(Base):
+    query_string = sa.Column(sa.String, nullable=False, unique=False)
+    groups = relationship("Group", secondary="group_filters", back_populates="filters")
+
+
+GroupFilter = join_model("group_filters", Group, Filter)
+
+Candidate = join_model("candidates", Filter, Obj)
+
+
+def candidate_is_owned_by(self, user_or_token):
+    return bool(set(self.filter.groups) & set(user_or_token.groups))
+
+
+Candidate.is_owned_by = candidate_is_owned_by
+
+
+Source = join_model("sources", Group, Obj)
 """User.sources defines the logic for whether a user has access to a source;
    if this gets more complicated it should become a function/`hybrid_property`
    rather than a `relationship`.
 """
-User.sources = relationship('Source', backref='users',
-                            secondary='join(Group, group_sources).join(group_users)',
+Source.saved_by_id = sa.Column(sa.ForeignKey("users.id"), nullable=True, unique=False)
```

Comment:
Can we do this with a mix-in class?

Reply:
Isn't a join table the appropriate thing here? This is just a join on Obj and Group so there's just obj_id and group_id coumns, then adding a few extra columns

Reply:
I don't think we want to duplicate any columns from Obj here.

Reply:
`Source` and `Candidate` should be join tables

Reply:
I think that fits my expectation too.